File: programmers/level2/조이스틱.py
def solution(name):
    change_needed = [0]*len(name)
    for idx, ch in enumerate(name):
        change_needed[idx] = min(ord(ch)-ord('A'), ord('Z')-ord(ch)+1)
    total = sum(change_needed)
    idx = ans = 0
    while total:
        ans += change_needed[idx]
        total -= change_needed[idx]
        if total == 0: return ans
        change_needed[idx] = 0
        
        left_midx = right_midx = 1
        while change_needed[idx-left_midx] == 0:
            left_midx += 1
        while idx+right_midx < len(change_needed) and change_needed[idx+right_midx] == 0:
            right_midx += 1
        if left_midx < right_midx: #가까운 거리로 이동
            idx -= left_midx
            ans += left_midx
        else:
            idx += right_midx
            ans += right_midx

File: programmers/level2/조이스틱.java
import java.util.*;

class Solution {
    public int solution(String name) {
        int answer = 0;
        int[] changeNeed = new int[name.length()];
        int total = 0;
        int idx = 0;
        for(int i=0; i<name.length(); i++) {
            changeNeed[i] = getChangeNum(name.charAt(i));
            total += changeNeed[i];
        }
        int deb = 0;
        while(total!=0) {
            int lMv = 0;
            int rMv = 1;
            if(changeNeed[idx]!=0) {
                answer += changeNeed[idx];
                total -= changeNeed[idx];
                changeNeed[idx] = 0;
            }
            if(total==0) break;
            int temp_idx = idx;
            while(changeNeed[temp_idx-lMv] == 0) {
                lMv++;
                if(temp_idx-lMv < 0) temp_idx += name.length();
            }
            while(idx+rMv < name.length() && changeNeed[idx+rMv] == 0)  rMv++;
            if(idx+rMv == name.length() || lMv < rMv) {
                idx = idx - lMv;
                if(idx<0) idx += name.length();
                answer += lMv;
            } else {
                idx = idx + rMv;
                answer += rMv;
            }
        }
        return answer;
    }
    
    public int getChangeNum(char ch) {
        return Math.min(ch-'A', 1+'Z'-ch);
    }
}